modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotWithMetastorageTest.java

```diff
@@ -0,0 +1,221 @@
+        AtomicBoolean stop = new AtomicBoolean();
+        CountDownLatch latch = new CountDownLatch(1);
+
+        IgniteEx ignite = startGridsWithCache(2, dfltCacheCfg, CACHE_KEYS_RANGE);
+
+        IgniteInternalFuture<?> updFut = GridTestUtils.runMultiThreadedAsync(() -> {
+            while (!Thread.currentThread().isInterrupted() && !stop.get()) {
+                try {
+                    ignite.context().distributedMetastorage().write(SNAPSHOT_PREFIX + keyCnt.getAndIncrement(), "value");
+                }
+                catch (IgniteCheckedException e) {
+                    throw new IgniteException(e);
+                }
+            }
+        }, 3, "dms-updater");
+
+        DmsDataWriterWorker worker = GridTestUtils.getFieldValue(ignite.context().distributedMetastorage(),
+            DistributedMetaStorageImpl.class, "worker");
+        LinkedBlockingQueue<RunnableFuture<?>> queue = GridTestUtils.getFieldValue(worker, DmsDataWriterWorker.class,
+            "updateQueue");
+
+        RunnableFuture<?> testTask = new FutureTask<>(() -> {
+            U.await(latch);
+
+            return null;
+        });
+
+        queue.offer(testTask);
+
+        assertTrue(GridTestUtils.waitForCondition(() -> queue.size() > 10, getTestTimeout()));
+
+        ignite.context().cache().context().exchange()
+            .registerExchangeAwareComponent(new PartitionsExchangeAware() {
+                /** {@inheritDoc} */
+                @Override public void onInitAfterTopologyLock(GridDhtPartitionsExchangeFuture fut) {
+                    latch.countDown();
+                }
+            });
+
+        ignite.snapshot().createSnapshot(SNAPSHOT_NAME).get();
+
+        stop.set(true);
+        updFut.get();
+
+        stopAllGrids();
+
+        IgniteEx snp0 = startGridsFromSnapshot(Collections.singleton(0),
+            cfg -> resolveSnapshotWorkDirectory(cfg).getAbsolutePath(), SNAPSHOT_NAME, false);
+
+        Set<String> allKeys = new HashSet<>();
+        snp0.context().distributedMetastorage().iterate(SNAPSHOT_PREFIX, (key, val) -> allKeys.add(key));
+
+        stopGrid(0);
+
+        IgniteEx snp1 = startGridsFromSnapshot(Collections.singleton(1),
+            cfg -> resolveSnapshotWorkDirectory(cfg).getAbsolutePath(), SNAPSHOT_NAME, false);
+
+        // Iterator reads keys from the node heap map.
+        snp1.context().distributedMetastorage()
+            .iterate(SNAPSHOT_PREFIX, new BiConsumer<String, Serializable>() {
+                @Override public void accept(String key, Serializable value) {
+                    try {
+                        assertTrue("Keys must be the same on all nodes [key=" + key + ", all=" + allKeys + ']',
+                            allKeys.contains(key));
+                    }
+                    catch (Throwable t) {
+                        fail("Exception reading metastorage: " + t.getMessage());
+                    }
+                }
+            });
```

Comment:
```suggestion
        Function<IgniteConfiguration, String> pathProv = cfg -> resolveSnapshotWorkDirectory(cfg).getAbsolutePath();
        Set<String> keySet0 = new TreeSet<>();
        Set<String> keySet1 = new TreeSet<>();

        IgniteEx snp0 = startGridsFromSnapshot(Collections.singleton(0), pathProv, SNAPSHOT_NAME, false);
        snp0.context().distributedMetastorage().iterate(SNAPSHOT_PREFIX, (key, val) -> keySet0.add(key));

        stopGrid(0);

        IgniteEx snp1 = startGridsFromSnapshot(Collections.singleton(1), pathProv, SNAPSHOT_NAME, false);
        snp1.context().distributedMetastorage().iterate(SNAPSHOT_PREFIX, (key, val) -> keySet1.add(key));

        assertEquals(keySet0, keySet1);
```

Reply:
TreeSet (in suggested code) required only for the output eg
``
java.lang.AssertionError: Keys must be the same on all nodes 
Expected :[SNAPSHOT_PREFIX_0, SNAPSHOT_PREFIX_1, SNAPSHOT_PREFIX_10, SNAPSHOT_PREFIX_11, SNAPSHOT_PREFIX_12, SNAPSHOT_PREFIX_2, SNAPSHOT_PREFIX_3, SNAPSHOT_PREFIX_4, SNAPSHOT_PREFIX_5, SNAPSHOT_PREFIX_6, SNAPSHOT_PREFIX_7, SNAPSHOT_PREFIX_8, SNAPSHOT_PREFIX_9]
Actual   :[SNAPSHOT_PREFIX_0, SNAPSHOT_PREFIX_1, SNAPSHOT_PREFIX_10, SNAPSHOT_PREFIX_100, SNAPSHOT_PREFIX_101, SNAPSHOT_PREFIX_102, SNAPSHOT_PREFIX_103, SNAPSHOT_PREFIX_104, SNAPSHOT_PREFIX_105, SNAPSHOT_PREFIX_106, SNAPSHOT_PREFIX_107, SNAPSHOT_PREFIX_108, SNAPSHOT_PREFIX_109, SNAPSHOT_PREFIX_11, SNAPSHOT_PREFIX_110, SNAPSHOT_PREFIX_111, SNAPSHOT_PREFIX_112, SNAPSHOT_PREFIX_113, SNAPSHOT_PREFIX_114, SNAPSHOT_PREFIX_115, SNAPSHOT_PREFIX_116, SNAPSHOT_PREFIX_117, SNAPSHOT_PREFIX_118, SNAPSHOT_PREFIX_119, SNAPSHOT_PREFIX_12, SNAPSHOT_PREFIX_120, SNAPSHOT_PREFIX_121, SNAPSHOT_PREFIX_122, SNAPSHOT_PREFIX_123, SNAPSHOT_PREFIX_124, SNAPSHOT_PREFIX_125, SNAPSHOT_PREFIX_126, SNAPSHOT_PREFIX_127, SNAPSHOT_PREFIX_128, SNAPSHOT_PREFIX_129, SNAPSHOT_PREFIX_13, SNAPSHOT_PREFIX_130, SNAPSHOT_PREFIX_131, SNAPSHOT_PREFIX_132, SNAPSHOT_PREFIX_133, SNAPSHOT_PREFIX_134, SNAPSHOT_PREFIX_135, SNAPSHOT_PREFIX_136, SNAPSHOT_PREFIX_137, SNAPSHOT_PREFIX_138, SNAPSHOT_PREFIX_139, SNAPSHOT_PREFIX_14, SNAPSHOT_PREFIX_140, SNAPSHOT_PREFIX_141, SNAPSHOT_PREFIX_142, SNAPSHOT_PREFIX_143, SNAPSHOT_PREFIX_144, SNAPSHOT_PREFIX_145, SNAPSHOT_PREFIX_146, SNAPSHOT_PREFIX_147, SNAPSHOT_PREFIX_148, SNAPSHOT_PREFIX_149, SNAPSHOT_PREFIX_15, SNAPSHOT_PREFIX_150, SNAPSHOT_PREFIX_151, SNAPSHOT_PREFIX_152, SNAPSHOT_PREFIX_153, SNAPSHOT_PREFIX_154, SNAPSHOT_PREFIX_155, SNAPSHOT_PREFIX_156, SNAPSHOT_PREFIX_157, SNAPSHOT_PREFIX_158, SNAPSHOT_PREFIX_159, SNAPSHOT_PREFIX_16, SNAPSHOT_PREFIX_160, SNAPSHOT_PREFIX_161, SNAPSHOT_PREFIX_162, SNAPSHOT_PREFIX_163, SNAPSHOT_PREFIX_164, SNAPSHOT_PREFIX_165, SNAPSHOT_PREFIX_166, SNAPSHOT_PREFIX_167, SNAPSHOT_PREFIX_168, SNAPSHOT_PREFIX_169, SNAPSHOT_PREFIX_17, SNAPSHOT_PREFIX_170, SNAPSHOT_PREFIX_18, SNAPSHOT_PREFIX_19, SNAPSHOT_PREFIX_2, SNAPSHOT_PREFIX_20, SNAPSHOT_PREFIX_21, SNAPSHOT_PREFIX_22, SNAPSHOT_PREFIX_23, SNAPSHOT_PREFIX_24, SNAPSHOT_PREFIX_25, SNAPSHOT_PREFIX_26, SNAPSHOT_PREFIX_27, SNAPSHOT_PREFIX_28, SNAPSHOT_PREFIX_29, SNAPSHOT_PREFIX_3, SNAPSHOT_PREFIX_30, SNAPSHOT_PREFIX_31, SNAPSHOT_PREFIX_32, SNAPSHOT_PREFIX_33, SNAPSHOT_PREFIX_34, SNAPSHOT_PREFIX_35, SNAPSHOT_PREFIX_36, SNAPSHOT_PREFIX_37, SNAPSHOT_PREFIX_38, SNAPSHOT_PREFIX_39, SNAPSHOT_PREFIX_4, SNAPSHOT_PREFIX_40, SNAPSHOT_PREFIX_41, SNAPSHOT_PREFIX_42, SNAPSHOT_PREFIX_43, SNAPSHOT_PREFIX_44, SNAPSHOT_PREFIX_45, SNAPSHOT_PREFIX_46, SNAPSHOT_PREFIX_47, SNAPSHOT_PREFIX_48, SNAPSHOT_PREFIX_49, SNAPSHOT_PREFIX_5, SNAPSHOT_PREFIX_50, SNAPSHOT_PREFIX_51, SNAPSHOT_PREFIX_52, SNAPSHOT_PREFIX_53, SNAPSHOT_PREFIX_54, SNAPSHOT_PREFIX_55, SNAPSHOT_PREFIX_56, SNAPSHOT_PREFIX_57, SNAPSHOT_PREFIX_58, SNAPSHOT_PREFIX_59, SNAPSHOT_PREFIX_6, SNAPSHOT_PREFIX_60, SNAPSHOT_PREFIX_61, SNAPSHOT_PREFIX_62, SNAPSHOT_PREFIX_63, SNAPSHOT_PREFIX_64, SNAPSHOT_PREFIX_65, SNAPSHOT_PREFIX_66, SNAPSHOT_PREFIX_67, SNAPSHOT_PREFIX_68, SNAPSHOT_PREFIX_69, SNAPSHOT_PREFIX_7, SNAPSHOT_PREFIX_70, SNAPSHOT_PREFIX_71, SNAPSHOT_PREFIX_72, SNAPSHOT_PREFIX_73, SNAPSHOT_PREFIX_74, SNAPSHOT_PREFIX_75, SNAPSHOT_PREFIX_76, SNAPSHOT_PREFIX_77, SNAPSHOT_PREFIX_78, SNAPSHOT_PREFIX_79, SNAPSHOT_PREFIX_8, SNAPSHOT_PREFIX_80, SNAPSHOT_PREFIX_81, SNAPSHOT_PREFIX_82, SNAPSHOT_PREFIX_83, SNAPSHOT_PREFIX_84, SNAPSHOT_PREFIX_85, SNAPSHOT_PREFIX_86, SNAPSHOT_PREFIX_87, SNAPSHOT_PREFIX_88, SNAPSHOT_PREFIX_89, SNAPSHOT_PREFIX_9, SNAPSHOT_PREFIX_90, SNAPSHOT_PREFIX_91, SNAPSHOT_PREFIX_92, SNAPSHOT_PREFIX_93, SNAPSHOT_PREFIX_94, SNAPSHOT_PREFIX_95, SNAPSHOT_PREFIX_96, SNAPSHOT_PREFIX_97, SNAPSHOT_PREFIX_98, SNAPSHOT_PREFIX_99]
<Click to see difference>
```